symfit/core/fit.py

```diff
@@ -1009,169 +1010,78 @@ class HasCovarianceMatrix(TakesData):
         except np.linalg.LinAlgError:
             return None
 
-    def _reduced_residual_ss(self, best_fit_params, flatten=False):
-        """
-        Calculate the residual Sum of Squares divided by the d.o.f..
-
-        :param best_fit_params: ``dict`` of best fit parameters as given by .best_fit_params()
-        :param flatten: when `True`, return the total sum of squares (SS).
-            If `False`, return the componentwise SS.
-        :return: The reduced residual sum of squares.
-        """
-        # popt = [best_fit_params[p.name] for p in self.model.params]
-        # Rescale the covariance matrix with the residual variance
-        if isinstance(self.objective, (VectorLeastSquares, LeastSquares)):
-            ss_res = self.objective(flatten_components=flatten,
-                                    **key2str(best_fit_params))
-        else:
-            ss_res = self.objective(**key2str(best_fit_params))
-
-        if isinstance(self.objective, VectorLeastSquares):
-            ss_res = np.sum(ss_res**2)
-
-        degrees_of_freedom = 0 if flatten else []
-        for data in self.dependent_data.values():
-            if flatten:
-                if data is not None:
-                    degrees_of_freedom += np.product(data.shape)
+        if isinstance(objective, LogLikelihood):
+            # Loglikelihood is a special case that needs to be considered
+            # separately, see #138.
+            return hess_inv
+        elif isinstance(objective, LeastSquares):
+            # Calculate the covariance for a least squares method.
+            # https://www8.cs.umu.se/kurser/5DA001/HT07/lectures/lsq-handouts.pdf
+            # Residual sum of squares
+            rss = objective(**key2str(best_fit_params))
+            # Degrees of freedom
+            raw_dof = np.sum([np.product(shape) for shape in self.data_shapes[1]])
+            dof = raw_dof - len(self.model.params)
+            if self.absolute_sigma:
+                # When interpreting as measurement error, we do not rescale.
+                s2 = 1
             else:
-                if data is not None:
-                    degrees_of_freedom.append(np.product(data.shape))
-                    # degrees_of_freedom = np.product(data.shape) - len(popt)
-                    # break
-                else: # the correspoding component in ss_res will be 0 so it is ok to add any non-zero number.
-                    degrees_of_freedom.append(len(best_fit_params) + 1)
-        degrees_of_freedom = np.array(degrees_of_freedom)
-        s_sq = ss_res / (degrees_of_freedom - len(best_fit_params))
-        # s_sq = ss_res / degrees_of_freedom
-
-        return s_sq
+                s2 = rss / dof
+            # Divide by two because the source uses different normalization
+            cov_mat = 2 * s2 * hess_inv
+            return cov_mat
+        else:
+            # We do not know how to handle with this situation.
+            return None
```

Comment:
This If/elif totem with isinstance calls probably means objectives should take care of calculating the covariance matrix